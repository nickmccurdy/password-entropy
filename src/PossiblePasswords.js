import { faList, faRandom } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import {
  Card,
  CardBody,
  FormGroup,
  FormText,
  Label,
  Progress,
} from 'reactstrap'

export default function PossiblePasswords({
  possiblePasswords,
  approximate,
  entropyBits,
  entropyTip: { strength },
}) {
  return (
    <Card>
      <CardBody>
        <FormGroup>
          <Label>
            <FontAwesomeIcon icon={faList} /> Entropy
          </Label>

          <FormText>
            <Progress max={128} value={entropyBits}>
              {entropyBits.toFixed(2)} bits ({strength})
            </Progress>
          </FormText>
        </FormGroup>

        <FormGroup>
          <Label>
            <FontAwesomeIcon icon={faRandom} /> Possible Passwords
          </Label>

          <FormText>
            {approximate && '~ '} {possiblePasswords}
          </FormText>
        </FormGroup>
      </CardBody>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired,
    style: PropTypes.string.isRequired,
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired,
}
