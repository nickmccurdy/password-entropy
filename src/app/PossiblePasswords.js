import { Card, CardText, LinearProgress } from 'material-ui'
import { colors } from 'material-ui/styles'
import PropTypes from 'prop-types'
import { always, filter, findLast, identity, keys, map, pipe, sum } from 'ramda'
import React from 'react'
import Icon from 'react-fa'
import { connect } from 'react-redux'

import { dicewareWords } from '../options/DicewareOptions'
import { toggles } from '../options/GenericOptions'

const entropyTips = [
  {
    color: colors.red500,
    minimum: 0,
    strength: 'Very Weak'
  },
  {
    color: colors.yellow500,
    minimum: 32,
    strength: 'Weak'
  },
  {
    color: colors.blue500,
    minimum: 64,
    strength: 'Strong'
  },
  {
    color: colors.green500,
    minimum: 128,
    strength: 'Very Strong'
  }
]

export function PossiblePasswords({
  possiblePasswords, approximate, entropyBits, entropyTip: { color, strength }
}) {
  return (
    <Card>
      <CardText>
        <h3><Icon name="list" /> Entropy</h3>
        {entropyBits.toFixed(2)} bits ({strength})
        <LinearProgress color={color} max={128} mode="determinate" value={entropyBits} />

        <h3><Icon name="random" /> Possible Passwords</h3>
        {approximate && '~ '}{possiblePasswords.toLocaleString()}
      </CardText>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    color: PropTypes.string.isRequired,
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired
}

const possibleItems = {
  generic: pipe(
    filter(identity),
    keys,
    map(key => toggles[key].possibleItems),
    sum
  ),
  diceware: always(dicewareWords),
  custom: identity
}

export function mapStateToProps({ length, options, optionsKey }) {
  const possiblePasswords = possibleItems[optionsKey](options[optionsKey]) ** length
  const entropyBits = Math.max(0, Math.log2(possiblePasswords))

  return {
    approximate: possiblePasswords > Number.MAX_SAFE_INTEGER,
    entropyBits,
    entropyTip: findLast(tip => entropyBits >= tip.minimum, entropyTips),
    possiblePasswords
  }
}

export default connect(mapStateToProps)(PossiblePasswords)
