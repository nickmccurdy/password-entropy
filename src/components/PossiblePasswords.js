import { faList, faRandom } from '@fortawesome/fontawesome-free-solid'
import Icon from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import React from 'react'
import {
  Card,
  CardBody,
  FormGroup,
  FormText,
  Label,
  Progress
} from 'reactstrap'

export default function PossiblePasswords({
  possiblePasswords,
  approximate,
  entropyBits,
  entropyTip: { strength, style }
}) {
  return (
    <Card>
      <CardBody>
        <FormGroup>
          <Label>
            <Icon icon={faList} /> Entropy
          </Label>

          <FormText>
            <Progress max={128} value={entropyBits}>
              {entropyBits.toFixed(2)} bits ({strength})
            </Progress>
          </FormText>
        </FormGroup>

        <FormGroup>
          <Label>
            <Icon icon={faRandom} /> PossiblePasswords
          </Label>

          <FormText>
            {approximate && '~ '} {possiblePasswords.toLocaleString()}
          </FormText>
        </FormGroup>
      </CardBody>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired,
    style: PropTypes.string.isRequired
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired
}
