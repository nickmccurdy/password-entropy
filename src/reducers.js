import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
