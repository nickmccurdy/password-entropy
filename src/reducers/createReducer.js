import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
