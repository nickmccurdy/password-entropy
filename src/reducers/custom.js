import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
