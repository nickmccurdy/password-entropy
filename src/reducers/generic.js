import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
