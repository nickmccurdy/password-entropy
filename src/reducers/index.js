import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
