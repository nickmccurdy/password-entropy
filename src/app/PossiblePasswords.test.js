import { shallow } from 'enzyme'
import { PossiblePasswords, mapStateToProps } from './PossiblePasswords'
import { assoc } from 'ramda'
import React from 'react'
import reducers from '../reducers'
import { createStore } from 'redux'

const getWrapper = length => shallow(<PossiblePasswords length={length} possibleItems={2}/>)

it('renders without possibleItems', () => {
  expect(shallow(<PossiblePasswords length={1}/>)).toMatchSnapshot()
})

it('renders with invalid possibleItems', () => {
  expect(shallow(<PossiblePasswords length={1} possibleItems={0}/>)).toMatchSnapshot()
})

it('renders with invalid or very weak possiblePasswords', () => {
  expect(getWrapper(0)).toMatchSnapshot()
})

it('renders with weak possiblePasswords', () => {
  expect(getWrapper(32)).toMatchSnapshot()
})

it('renders with strong possiblePasswords', () => {
  expect(getWrapper(64)).toMatchSnapshot()
})

it('renders with very strong possiblePasswords', () => {
  expect(getWrapper(128)).toMatchSnapshot()
})

test('mapStateToProps', () => {
  function mapStateToPropsWithIndex (optionsIndex) {
    const state = createStore(reducers).getState()
    return mapStateToProps({ ...state, optionsIndex })
  }

  expect(mapStateToPropsWithIndex(0)).toEqual({ possibleItems: 7776 })
  expect(mapStateToPropsWithIndex(1)).toEqual({ possibleItems: 1 })
  expect(mapStateToPropsWithIndex(2)).toEqual({ possibleItems: 70 })
})
