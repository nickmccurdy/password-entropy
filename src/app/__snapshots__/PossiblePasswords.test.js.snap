// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PossiblePasswords approximate 1`] = `
<Panel
  bsClass="panel"
  bsStyle="default"
  defaultExpanded={false}
>
  <FormGroup
    icon="list"
    id="entropy"
    label="Entropy"
  >
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="info"
      isChild={false}
      label="77.55 bits (Strong)"
      max={128}
      min={0}
      now={77.54887502163469}
      srOnly={false}
      striped={false}
    />
  </FormGroup>
  <FormGroup
    icon="random"
    id="possiblePasswords"
    label="PossiblePasswords"
  >
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      ~ 
      221,073,919,720,733,000,000,000
    </FormControlStatic>
  </FormGroup>
</Panel>
`;

exports[`PossiblePasswords exact 1`] = `
<Panel
  bsClass="panel"
  bsStyle="default"
  defaultExpanded={false}
>
  <FormGroup
    icon="list"
    id="entropy"
    label="Entropy"
  >
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="info"
      isChild={false}
      label="77.55 bits (Strong)"
      max={128}
      min={0}
      now={77.54887502163469}
      srOnly={false}
      striped={false}
    />
  </FormGroup>
  <FormGroup
    icon="random"
    id="possiblePasswords"
    label="PossiblePasswords"
  >
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      221,073,919,720,733,000,000,000
    </FormControlStatic>
  </FormGroup>
</Panel>
`;

exports[`mapStateToProps custom 1`] = `
Object {
  "approximate": false,
  "entropyBits": 0,
  "entropyTip": Object {
    "minimum": 0,
    "strength": "Very Weak",
    "style": "danger",
  },
  "possiblePasswords": 1,
}
`;

exports[`mapStateToProps custom with invalid possibleItems 1`] = `
Object {
  "approximate": false,
  "entropyBits": 0,
  "entropyTip": Object {
    "minimum": 0,
    "strength": "Very Weak",
    "style": "danger",
  },
  "possiblePasswords": 0,
}
`;

exports[`mapStateToProps custom with strong possiblePasswords 1`] = `
Object {
  "approximate": true,
  "entropyBits": 64,
  "entropyTip": Object {
    "minimum": 64,
    "strength": "Strong",
    "style": "info",
  },
  "possiblePasswords": 18446744073709552000,
}
`;

exports[`mapStateToProps custom with very strong possiblePasswords 1`] = `
Object {
  "approximate": true,
  "entropyBits": 128,
  "entropyTip": Object {
    "minimum": 128,
    "strength": "Very Strong",
    "style": "success",
  },
  "possiblePasswords": 3.402823669209385e+38,
}
`;

exports[`mapStateToProps custom with weak possiblePasswords 1`] = `
Object {
  "approximate": false,
  "entropyBits": 32,
  "entropyTip": Object {
    "minimum": 32,
    "strength": "Weak",
    "style": "warning",
  },
  "possiblePasswords": 4294967296,
}
`;

exports[`mapStateToProps diceware 1`] = `
Object {
  "approximate": true,
  "entropyBits": 77.54887502163469,
  "entropyTip": Object {
    "minimum": 64,
    "strength": "Strong",
    "style": "info",
  },
  "possiblePasswords": 2.2107391972073336e+23,
}
`;

exports[`mapStateToProps generic 1`] = `
Object {
  "approximate": false,
  "entropyBits": 36.775698101669796,
  "entropyTip": Object {
    "minimum": 32,
    "strength": "Weak",
    "style": "warning",
  },
  "possiblePasswords": 117649000000,
}
`;
